comiccrawler/core/__init__.py

```diff
@@ -611,40 +611,64 @@ def analyze(mission):
 		download_ch.pub("ANALYZE_FINISHED", mission)
 		
 class BatchAnalyzer:
-	def __init__(self, missions):
+	def __init__(
+		self,
+		gen_missions,
+		stop_on_error=True,
+		done_item=None,
+		done=None
+	):
 		self.thread = Worker(self.analyze)
-		self.missions = deque(missions)
-		self.lock = RLock()
+		self.gen_missions = gen_missions
+		self.done = done
+		self.done_item = done_item
+		self.stop_on_error = stop_on_error
+		self.cooldown = {}
+		self.last_err = None
 		
 	def start(self):
 		self.thread.start()
 		return self
 		
 	def stop(self):
 		self.thread.stop()
+		return self
+		
+	def get_cooldown(self, mission):
+		if not hasattr(mission.module, "rest_analyze"):
+			return 0
+		pre_ts = self.cooldown.get(mission.module.name)
+		if pre_ts is None:
+			return 0
+		cooldown = mission.module.rest_analyze - (time() - pre_ts)
+		return cooldown if cooldown > 0 else 0
 		
 	def analyze(self):
-		err = None
-		
-		while self.missions:
-			mission = None
-			with self.lock:
-				mission = self.missions.popleft()
-			with mission.load_episode():
-				try:
-					Analyzer(mission).analyze()
-				except BaseException as _err:
-					with self.lock:
-						self.missions.appendleft(mission)
-					err = _err
+		self.last_err = None
+		try:
+			self.do_analyze()
+		finally:
+			if self.done:
+				self.done(self.last_err)
+			
+	def do_analyze(self):
+		for mission in self.gen_missions:
+			try:
+				sleep(self.get_cooldown(mission))
```

Comment:
Currently, the list is updated before sleep (in the generator). It would be better if the list is updated before analyzer, after sleep (a new hook?).

Reply:
Or update the list in done_item?